staging/vhost-device-can/README.md

```diff
@@ -0,0 +1,114 @@
+# vhost-device-can - CAN emulation backend daemon
+
+## Description
+This program is a vhost-user backend that emulates a VirtIO CAN device.
+The device's binary takes two (2) parameters:  a socket, a 'can-devices' list.
+The socket is commonly used across all vhost-devices to communicate with
+the vhost-user frontend device. The 'can-devices' represents a list of
+CAN/FD devices appears in the host system. This list includes *can_input* and
+*can_output* devices which vhost-device-can will forward messages to and from
+the frontend side.
+
+This program is tested with QEMU's `vhost-user-device-pci` device.
+Examples' section below.
+
+## Synopsis
+```
+**vhost-device-can** [*OPTIONS*]
+````
+
+## Options
+
+.. program:: vhost-device-can
+
+.. option:: -h, --help
+
+  Print help.
+
+.. option:: -s, --socket-path=PATH
+
+  Location of vhost-user Unix domain sockets, this path will be suffixed with
+  0,1,2..socket_count-1.
+
+.. option:: -c, --socket-count=INT
+
+  Number of guests (sockets) to attach to, default set to 1.
+
+.. option:: -d, --can-devices='CAN/FD interfaces'
+
+  CAN/FD device list at the host OS in the format:
+      <can-in_1>:<can_out_1> [<can_in_2>:<can_out_2>] ... [<can_in_N>:<can_out_N>]
+
+  Note: Where N is equal with the number provided via *socket_count* parameter.
+
+      Example: --can-devices "vcan0:vcan1 vcan2:vcan3"
```

Comment:
I am still trying to get familiar with CAN. Why is the parameter organized in couples? Does each couple represent a single CAN interface that is exposed to guest? Is not possible to use an interface in the host for both send and receive msgs? Why do they have to be different interfaces for input and output? Thanks.

Reply:
> I am still trying to get familiar with CAN. Why is the `--can-devices` parameter organized in couples? Does each couple represent a single CAN interface that is exposed to guest? Is it not possible to use an interface in the host for both send and receive msgs? Why do they have to be different interfaces for input and output? Thanks.

Oh, I think the number of couples is equal to the socket_count-1.

Reply:
In most cases, we need to send and receive messages to/from the same interface so the couple will be `--can-devices "can0:can0"`. But there are cases, where we might have 2 CAN/FD available channels (connected in the same CAN bus) and we want/need to use *can0* as receiver and *can1* as sender (`--can-devices "can0:can1"`). To cover scenarios like this, I created this couple structure in order the user to have that freedom.

>Oh, I think the number of couples is equal to socket_count, but I still do not understand why I need different interfaces for input and output.

Thanks for that, I just update it to make clear that N is the number of sockets and couples.

Reply:
> In most cases, we need to send and receive messages to/from the same interface so the couple will be `--can-devices "can0:can0"`. But there are cases, where we might have 2 CAN/FD available channels (connected in the same CAN bus) and we want/need to use _can0_ as receiver and _can1_ as sender (`--can-devices "can0:can1"`). To cover scenarios like this, I created this couple structure in order the user to have that freedom.

Cool, I think it could be nice to document that somewhere.

Reply:
Good point, thanks, I have just added it in the readme file. I might revise it in the near future.

Reply:
Hello Matias, I have simplified the logic a bit. The current implementation eliminates the "tuples" logic `--can-devices "can0:can1"`. Instead uses `--can-devices "can0"` to relate the "can0" host interface with the produced socket path (`/tmp/sock0`). Then QEMU VM can use this host interface for both transmission and receiption of CAN/FD frames. In the odd case which I was describing in the past:
> there are cases, where we might have 2 CAN/FD available channels (connected in the same CAN bus) and we want/need to use can0 as receiver and can1 as sender

This can be addressed by running:
```
vhost-device-can --socket-path /tmp/sock --socket-count 2 --can-devices "can0 can1"
```
And eventually running th QEMU VM with two vhost-user-devices as:
```
host# qemu-system                                         \
    -m 4096                                               \
    -numa node,memdev=mem1                                 \
    -object memory-backend-memfd,id=mem1,size=2G,share=on  \
    -chardev socket,path=/tmp/can.sock,id=chcan1            \
    -device vhost-user-can-pci,chardev=chcan1,id=can1        \
    -numa node,memdev=mem2                                 \
    -object memory-backend-memfd,id=mem2,size=2G,share=on  \
    -chardev socket,path=/tmp/can.sock,id=chcan2            \
    -device vhost-user-can-pci,chardev=chcan2,id=can2        \
    ...
```

In this case, the guest will have two CAN/FD interfaces available and can use one of transmission and the other for receiption.